### ISSUE TITLE
Allow restarting the JustSayingBus once it is stopped

### ISSUE THREAD

**Issue by bjornsy-tl** [justeattakeaway/JustSaying#1155]
**Is your feature request related to a problem? Please describe.**
At the moment my team is using JustSaying v6, for several reasons we'd like to update to v7. It works well except for when we need to restart the bus after stopping listening (by cancelling the CancellationToken), due to switching between blue and green slices with deployments (i.e. we don't want the offline slice to keep listening and processing messages on an older version of the code, but when it's toggled to online, it needs to start listening straight away).

I believe this is due to the new JustSayingBus only having `StartAsync` which uses a [`_busStarted` boolean which is always `true`](https://github.com/justeattakeaway/JustSaying/blob/b30ec797d3bd672a4d875aa1051fac1f1924d34a/src/JustSaying/JustSayingBus.cs#L137) once initially started, and when `StartAsync` is called again the [if statement](https://github.com/justeattakeaway/JustSaying/blob/b30ec797d3bd672a4d875aa1051fac1f1924d34a/src/JustSaying/JustSayingBus.cs#L123) prevents it actually restarting.

**Describe the solution you'd like**
Perhaps a new method called `RestartAsync` that allows starting the bus when `_busStarted` is true and the same stopping token that was initially passed in has been cancelled. Open to discussions for this or alternatives, and opening a PR myself

**Describe alternatives you've considered**
A method that manually sets `_busStarted` to false. It doesn't feel as clean though

**Additional context**
Currently for v6 we call StartListening and StopListening as required
https://github.com/justeattakeaway/JustSaying/blob/c25443d907aa4d654ed8b7c1f8dcf2dd76542b8c/JustSaying/JustSayingFluently.cs#L155-L168

which allows restarting easily as it is not controlled by a CancellationToken
https://github.com/justeattakeaway/JustSaying/blob/c25443d907aa4d654ed8b7c1f8dcf2dd76542b8c/JustSaying/JustSayingBus.cs#L116-L155

**Comment by gkinsman** [comment]
Hi there! I might be able to provide some context on why this changed from v6 to v7, and a possible solution to your needs.

JustSaying v6 has a monolithic design in which there was a central message pump component ([SqsNotificationListener](https://github.com/justeattakeaway/JustSaying/blob/c25443d907aa4d654ed8b7c1f8dcf2dd76542b8c/JustSaying/AwsTools/MessageHandling/SqsNotificationListener.cs#L1)) that is responsible for downloading messages, throttling consumption, and dispatching to consumers. The lack of modularity and problems with the throttling mechanism meant that we couldn't implement correct backpressure so that JustSaying would stop downloading messages when it had already downloaded enough, which caused problems at scale. 

We opted to overhaul these internals and modularise, and an impact of that was that the thing that downloads messages was no longer also the thing that dispatches them (instead they are now connected by channels), so it was no longer possible to have a single choke point that can start and stop the bus as needed. The new design follows other middlewares like MassTransit and ASP.NET in being start-once. Once the bus is started, the application must be restarted to stop/start consumption.

On to a solution to your problem! If I understand correctly, what you want is to be able to start/stop the bus from downloading messages from SQS at will so that you can swap between active deployments. 
I would start by looking at the MessageReceiveBuffer, which is responsible for downloading messages, and introducing some kind of global controller that can be injected via DI, and used to stop/start all receive buffers at once. The receive loop could check its status to determine if it should allow the loop to continue, or wait and start the loop again.

Something like:
```csharp
interface IReceiveController {
    void Stop();
    void Start();
}
```

I hope this helps, good luck!

**Comment by georgearnall** [comment]
Any news on when 7.0.2 will be released? We are using this package and weren't aware this was dropped in v7.

**Comment by slang25** [comment]
We've just released a `7.1.0` with this change, thanks for the contribution 🥳 
https://github.com/justeattakeaway/JustSaying/releases/tag/v7.1.0
https://www.nuget.org/packages/JustSaying/7.1.0
